datafusion/core/src/datasource/schema_adapter.rs

```diff
@@ -79,11 +87,17 @@ pub trait SchemaAdapter: Send + Sync {
     ) -> datafusion_common::Result<(Arc<dyn SchemaMapper>, Vec<usize>)>;
 }
 
-/// Maps, by casting or reordering columns from the file schema to the table
-/// schema.
+/// Maps, columns from the file schema to the table schema.
+///
+/// See [`DefaultSchemaAdapterFactory`] for more details and examples.
 pub trait SchemaMapper: Debug + Send + Sync {
-    /// Adapts a `RecordBatch` to match the `table_schema` using the stored
-    /// mapping and conversions.
+    /// Adapts a `RecordBatch` to match the `table_schema`
+    ///
+    /// # Errors:
+    ///
+    /// * If a column in the table schema is non-nullable but is not present
+    ///   in the file schema (i.e. it is missing), this method tries to fill it
+    ///   with nulls resulting in a schema error
```

Comment:
I think this is specific to the `DefaultSchemaAdapterFactory` and should be moved to there. You could imagine that someone writes a mapper that fills in missing columns with placeholder values or intra-/extrapolates the data.

Reply:
It is a good call -- tried to clarify in d966fd112